app/scripts/components/PromoTransactionsTable/PromoTransactionsTable.tsx

```diff
@@ -0,0 +1,223 @@
+import { PromoRegistration } from '../../hooks/usePromoRegistrationList';
+import classNames from 'classnames';
+import React, { useMemo } from 'react';
+import { OverlayTrigger, Popover } from 'react-bootstrap';
+import { RegistrationsPaidPopover } from '../RegistrationsPaidPopover/RegistrationsPaidPopover';
+
+export interface PromoTransactionsTableProps {
+  // The promo registrations to display
+  promoRegistrations: Array<PromoRegistration>;
+
+  // The id of the report that is currently being viewed, if any
+  currentReportId: string | null;
+
+  // The localized currency symbol
+  currencySymbol: string;
+
+  // Message to display when there aren't any promo transactions
+  emptyMessage?: string;
+
+  // Extra content to display in the table header
+  headerExtra?: React.ReactFragment;
+
+  // Convert a currency amount to a localized string
+  localizedCurrency: (amount: number) => string;
+
+  // The set of registrations that are selected and will be included in a new report
+  selectedRegistrations: Set<PromoRegistration>;
+
+  // Handle changes to the selected state of a promo transaction
+  setRegistrationSelected: (
+    promoRegistration: PromoRegistration,
+    selected: boolean,
+  ) => void;
+
+  // The title of the transactions table
+  title: string;
+
+  // Open the view payments modal
+  viewPayments: (registrationId: string) => void;
+}
+
+export const PromoTransactionsTable = ({
+  promoRegistrations,
+  currentReportId,
+  currencySymbol,
+  emptyMessage,
+  headerExtra,
+  localizedCurrency,
+  selectedRegistrations,
+  setRegistrationSelected,
+  title,
+  viewPayments,
+}: PromoTransactionsTableProps): JSX.Element => {
+  const getBalanceClassName = (balance: number): string => {
+    if (balance === 0) {
+      return 'btn-success';
+    } else if (balance < 0) {
+      return 'btn-danger';
+    } else {
+      return 'btn-default';
+    }
+  };
+
+  const rows = useMemo(
+    () =>
+      // Expand each registration item into one row per registrant in the registration
+      promoRegistrations.flatMap((promoRegistration) =>
+        promoRegistration.registration.groupRegistrants
+          // If the promotion doesn't apply to all registrations, filter out the non-primary registrants
+          .filter(
+            (registrant) =>
+              promoRegistration.promotion.applyToAllRegistrants ||
+              registrant.id ===
+                promoRegistration.registration.primaryRegistrantId,
+          )
+          .map((registrant) => ({
+            id: `${registrant.id}|${promoRegistration.promotion.id}`,
+            promoRegistration,
+            registrant,
+            ...promoRegistration,
+          })),
+      ),
+    [promoRegistrations],
```

Comment:
This shouldn't stop this from going live, but something to fix at a later date.

On re-renders, React is caching a new version of the `promoRegistrations` array in PromoUploadPage.tsx, which will trigger this `useMemo` to run the calulation again with the new `promoRegistrations` as it's a dependency.

The way to fix it is to add an `useMemo` to the creation of `promoRegistrations`, so it only runs the function again if data in it's dependency changes.

Reply:
I just fixed what I believe you're referring to.